# Step 1: Importing useful libraries and loading the data

To get started, we're going to bring in pandas, numpy, matplotlib, and seaborn.

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Let's load in the dataset, "kc_house_data.csv" into a pandas dataframe.

In [16]:
df = pd.read_csv('kc_house_data.csv')
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view     ...      grade  sqft_above  \
0      5650     1.0         NaN   0.0     ...          7        1180   
1      7242     2.0         0.0   0.0     ...          7        2170   
2     10000     1.0         0.0   0.0     ...          6         770   
3      5000     1.0         0.0   0.0     ...          7        1050   
4      8080     1.0         0.0   0.0     ...          8        1680   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

# Step 2:  Review and understand the data

The import looks like it went well, so from here we're going to work on taking a look at the overall dataset and just getting a feel for the categorical data that's stored within it so that we can better understand how to interpret it.

In [17]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.159700e+04  21597.000000  19221.000000  21534.000000  21597.000000   
mean   1.509941e+04      1.494096      0.007596      0.233863      3.409825   
std    4.141264e+04      0.539683      0.086825      0.765686      0.650546   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068500e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above      yr_built  yr_renovated       zipcode  \
count  21597.000000  21597.000000  21597.000000  17755.000000  21597.000000   
mean       7.657915   1788.596842   1970.999676     83.636778  98077.951845   
std        1.173200    827.759761     29.375234    399.946414     53.513072   
min        3.000000    370.000000   1900.000000      0.000000  98001.000000   
25%        7.000000   1190.000000   1951.000000      0.000000  98033.000000   
50%        7.000000   1560.000000   1975.000000      0.000000  98065.000000   
75%        8.000000   2210.000000   1997.000000      0.000000  98118.000000   
max       13.000000   9410.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000   21597.000000   21597.000000  
mean      47.560093   -122.213982    1986.620318   12758.283512  
std        0.138552      0.140724     685.230472   27274.441950  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471100   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.231000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

In [18]:
df.shape

(21597, 21)

Calling `df.shape` tells us right away that our dataset looks at ~21.6K individual listings and reviews them across 21 different categories. Combining this with the info we can see in `df.head()`, we can make several useful inferences:

* If the data is complete, every column should have 21,597 entries.
* We have a lot of categorical data, which means we'll need to find information on what these categories signify.
* There's an included column called 'ID', which we could use as the index if we so chose.

Flatiron modified this dataset, and they went ahead and included some descriptions. If these weren't provided, we'd have to go look at Kaggle to find our parent data set and its descriptors. Here is what the available descriptions say about our more obtuse categories:

* **sqft_living:**  represents the living room area.
* **waterfront:**  binary represntation as to whether the property has a waterfront view.
* **view:** n times the house has been viewed.
* **condition:** represents the overall condition of the property on a scale of 1 to 5.
* **grade:** overall grade given to the housing unit, based on King County grading system. Scale of 1 to 13.
* **sqft_above:** is the square footage of the house apart from the basement.
* **sqft_lot15:** is the square footage of the lot in 2015.

# Step 3: Let's get this cleaned up.

For our first test of completeness, we can check `df.info()` and `df.isna().sum()` to see what issues we can identify:

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
id               21597 non-null int64
date             21597 non-null object
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
dtypes: float64(8), int64(11), object(2)
memory usage: 3.5+ MB


In [20]:
df.isna().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

Already, we can tease out some issues:

* 'date' should be a 'datetime' type, not an 'object' type.

* 'yr_renovated' and 'yr_built' could be changed to datetime, but years can stand independently as number types - unlike month/day/year combinations.

* 'sqft_basement' should be a number type, not an 'object.'

* We've got missing data in three columns: 1) 'Waterfront,' 2) 'View,' and 3) 'Yr_renovated.' We need to reconcile this before we move forward.

Let's tackle that date column first, followed by our basement area column, and then let's make sure that proceeded according to plan.

In [21]:
df['date'] = pd.to_datetime(df.date)
df['sqft_basement'].apply(lambda x: float(x))
df.dtypes

ValueError: could not convert string to float: '?'

Looks like everything worked out for that date column, but our sqft_basement column looks like it's in trouble. Based on that error, it looks like it's got '?' in place for NaN. Let's check this assumption:

In [22]:
df.loc[df['sqft_basement'] == '?'].count()

id               454
date             454
price            454
bedrooms         454
bathrooms        454
sqft_living      454
sqft_lot         454
floors           454
waterfront       417
view             452
condition        454
grade            454
sqft_above       454
sqft_basement    454
yr_built         454
yr_renovated     366
zipcode          454
lat              454
long             454
sqft_living15    454
sqft_lot15       454
dtype: int64

So, it looks like we have 454 properties w/ an unlisted basement square footage given as '?'. This means that we now know that there are four columns which contain unknown/empty values, and not just the previous three.

## Let's Start Pruning

At this point, we've decided that we're going to have to make-do with what we have available, so let's start evaluating ways to clean up our data. For sure, if we can remove bad rows from our set, we'd like to do that first.

#### What's a bad row?

For our intents and purposes, a 'bad row' is going to be any place where we're going to have to estimate multiple  points of our missing data. Since there are only 20 columns (we're excluding the ID column), if we're estimating two points of data for that row, we're estimating 10% of the data that row can offer. It's not really the best place to find ourselves in.

However, by making sure that we only focus on rows where we're making multiple estimations, we should minimize the amount of rows that we're removing, and also reduce the risk of introducing unnecessary spread through estimation. Let's start by taking a look at removing rows for which we'd have to estimate 4 points of data:

In [23]:
df.loc[(df['yr_renovated'].isnull() == True) & (df['waterfront'].isnull() == True) & (df['sqft_basement'] == '?') & (df['view'].isnull() == True)]

Empty DataFrame
Columns: [id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

[0 rows x 21 columns]

Looks like that worked out:  there are no rows where we have to guess 4 points of data. Let's narrow it down to three columns. Which one would be a good column to ignore?

Since 'view' represents the number of viewings a house gets, this might be a good one to cull. Let's take a look at our 'view' column in more detail.

In [24]:
df['view'].unique(), df['view'].isnull().sum()

(array([ 0., nan,  3.,  4.,  2.,  1.]), 63)

In [25]:
df.groupby('view')['id'].count()

view
0.0    19422
1.0      330
2.0      957
3.0      508
4.0      317
Name: id, dtype: int64

Since 'view' represents only 63 misses for the total dataset, and on the whole 'view' skews heavily toward 0, it really isn't going to risk killing our model. At this point, it's safe to pare off 'view' and proceed with our more frequent missing values.

In [26]:
all_3 = df.loc[(df['yr_renovated'].isnull() == True) & (df['waterfront'].isnull() == True) & (df['sqft_basement'] == '?')]
all_3.head()

id       date      price  bedrooms  bathrooms  sqft_living  \
3375  3629921240 2014-07-28   970000.0         4       4.50         3890   
4913  9257900010 2015-04-22   499900.0         4       2.25         2360   
7070  2724079090 2015-01-05  1650000.0         4       3.25         3920   
7325  4310700020 2014-10-10   280000.0         3       1.00         1100   
8905   822059038 2014-07-31   290000.0         6       4.50         2810   

      sqft_lot  floors  waterfront  view     ...      grade  sqft_above  \
3375      5906     2.0         NaN   3.0     ...         11        3060   
4913      7650     1.0         NaN   0.0     ...          8        1640   
7070    881654     3.0         NaN   3.0     ...         11        3920   
7325      5132     1.0         NaN   0.0     ...          6         840   
8905     11214     1.0         NaN   0.0     ...          8        2010   

      sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
3375              ?     2004           NaN    98029  47.5426 -121.995   
4913              ?     1963           NaN    98155  47.7500 -122.292   
7070              ?     2002           NaN    98024  47.5385 -121.896   
7325              ?     1948           NaN    98103  47.7011 -122.336   
8905              ?     1958           NaN    98031  47.4045 -122.197   

      sqft_living15  sqft_lot15  
3375           4170        6052  
4913           2320       11060  
7070           2970      112384  
7325           1280        5132  
8905           1940        8349  

[5 rows x 21 columns]

So, if we do this, we're only risking killing off 5 rows total. Not bad. What if we go all the way and get rid of rows that have two or more missing criteria?

In [27]:
yr_basement = df.loc[(df['yr_renovated'].isnull() == True) & (df['sqft_basement'] == '?')]
yr_basement.shape

(88, 21)

In [28]:
yr_waterfront = df.loc[(df['yr_renovated'].isnull() == True) & (df['waterfront'].isnull() == True)]
yr_waterfront.shape

(430, 21)

In [29]:
basement_waterfront = df.loc[(df['sqft_basement'] == '?') & (df['waterfront'].isnull() == True)]
basement_waterfront.shape

(37, 21)

Altogether, it looks like we could throw away almost 600 rows of data, or about 3% of our total data. This is a pretty big shift to make that carries risks - so, I'm going to pull it all out and into a separate, modified dataframe.

In [30]:
drop_these_indices = np.concatenate((np.array(all_3.index), np.array(yr_basement.index), np.array(yr_waterfront.index), np.array(basement_waterfront.index)))
drop_these_indices.size

561

In [31]:
df_truncated = df.drop(drop_these_indices)
df_truncated.shape

(21054, 21)

In [32]:
df_truncated.loc[df_truncated['id'] == 822059038]

Empty DataFrame
Columns: [id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

[0 rows x 21 columns]

From here, we can see that our truncated dataframe has removed all instances that we described previously. Since we've taken care of rows that would require multiple guesses,  let's take a look at what's left to clean up. 

Again, we're going to query `df.isna().sum()` and then count the number of remaining rows that have our placeholder '?' value in sqft_basement.

In [33]:
df_truncated.isna().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       1915
view               61
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3330
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

In [34]:
df_truncated.loc[df_truncated['sqft_basement'] == '?'].count()

id               335
date             335
price            335
bedrooms         335
bathrooms        335
sqft_living      335
sqft_lot         335
floors           335
waterfront       335
view             333
condition        335
grade            335
sqft_above       335
sqft_basement    335
yr_built         335
yr_renovated     335
zipcode          335
lat              335
long             335
sqft_living15    335
sqft_lot15       335
dtype: int64

Let's tackle 'Waterfront' and 'Yr_renovated' and 'View' first.

In [35]:
df_truncated['waterfront'].describe()

count    19139.000000
mean         0.007628
std          0.087009
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: waterfront, dtype: float64

In [36]:
df_truncated['yr_renovated'].describe()

count    17724.000000
mean        83.558677
std        399.769843
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       2015.000000
Name: yr_renovated, dtype: float64

In [37]:
df_truncated['view'].describe()

count    20993.000000
mean         0.233221
std          0.764412
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          4.000000
Name: view, dtype: float64

For all three of these guys, it looks like their distribution heavily skews toward '0'. Standard practice is to replace an unknown value with the median for that column, which in this case is already '0'. So, assigning these values as '0' isn't likely to skew the dataset in any direction, and is already consistent with what the real data is telling us.

In [38]:
df_truncated['waterfront'].fillna(0, inplace = True)
df_truncated['yr_renovated'].fillna(0, inplace = True)
df_truncated['view'].fillna(0, inplace = True)
df_truncated.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

Alright, since that worked out, let's take a look at filling in that placeholder value with something more useful and converting that column into a number.

In [39]:
def kill_q(q):
    if q == '?':
        return 0
    else:
        return q
df_truncated['sqft_basement'] = df_truncated.sqft_basement.map(lambda x: kill_q(x))
df_truncated.loc[df_truncated['sqft_basement'] == '?'].count()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [40]:
df_truncated['sqft_basement'] = df_truncated.sqft_basement.map(lambda x: float(x))
df_truncated.dtypes

id                        int64
date             datetime64[ns]
price                   float64
bedrooms                  int64
bathrooms               float64
sqft_living               int64
sqft_lot                  int64
floors                  float64
waterfront              float64
view                    float64
condition                 int64
grade                     int64
sqft_above                int64
sqft_basement           float64
yr_built                  int64
yr_renovated            float64
zipcode                   int64
lat                     float64
long                    float64
sqft_living15             int64
sqft_lot15                int64
dtype: object

# Okay, that's it!

I've already saved this cleaned up dataset as a pkl file to use in our exploratory phase covered in the next journal.